ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/deduplication/DeduplicationPeriodService.scala

```diff
@@ -0,0 +1,70 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.ledger.participant.state.kvutils.deduplication
+
+import java.time.{Duration, Instant}
+
+import akka.stream.Materializer
+import com.daml.error.ContextualizedErrorLogger
+import com.daml.ledger.api.DeduplicationPeriod
+import com.daml.ledger.api.domain.ApplicationId
+import com.daml.lf.data.Ref
+import com.daml.logging.LoggingContext
+import com.daml.platform.server.api.validation.{DeduplicationPeriodValidator, ErrorFactories}
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class DeduplicationPeriodService(
+    converter: DeduplicationPeriodConverter,
+    validation: DeduplicationPeriodValidator,
+    errorFactories: ErrorFactories,
+) {
+
+  def supportedDeduplicationPeriod(
+      deduplicationPeriod: DeduplicationPeriod,
+      maxDeduplicationDuration: Duration,
+      applicationId: ApplicationId,
+      actAs: Set[Ref.Party],
+      submittedAt: Instant,
+  )(implicit
+      mat: Materializer,
+      ec: ExecutionContext,
+      loggingContext: LoggingContext,
+      contextualizedErrorLogger: ContextualizedErrorLogger,
+  ): Future[DeduplicationPeriod] = {
+    val validatedDeduplicationPeriod = deduplicationPeriod match {
+      case period: DeduplicationPeriod.DeduplicationDuration =>
+        Future.successful(validation.validate(period, maxDeduplicationDuration))
+      case DeduplicationPeriod.DeduplicationOffset(offset) =>
+        converter
+          .convertOffsetToDuration(offset.toHexString, applicationId, actAs, submittedAt)
+          .map(
+            _.fold(
+              {
+                case DeduplicationConversionFailure.CompletionAtOffsetNotFound |
```

Comment:
Is there any reason why we have so many errors if we don't use them here?

Reply:
Ideally, we would further propagate them on. But currently, we directly propagate gRPC exceptions. I hope to change that in the future